packages/xo-server/src/_dedupeUnmount.js

```diff
@@ -0,0 +1,45 @@
+import assert from 'assert'
+
+import ensureArray from './_ensureArray'
+import MultiKeyMap from './_MultiKeyMap'
+
+function State() {
+  this.i = 0
+  this.value = undefined
+}
+
+export const dedupeUnmount = (fn, keyFn) => {
+  const states = new MultiKeyMap()
+
+  return function() {
+    const keys = ensureArray(keyFn.apply(this, arguments))
+    let state = states.get(keys)
+    if (state === undefined) {
+      state = new State()
+      states.set(keys, state)
+
+      const mount = async () => {
+        try {
+          const value = await fn.apply(this, arguments)
+          return {
+            __proto__: value,
+            async unmount() {
+              assert(state.i > 0)
+              if (--state.i === 0) {
+                states.delete(keys)
+                await value.unmount()
+              }
+            },
+          }
+        } catch (error) {
+          states.delete(keys)
+          throw error
+        }
+      }
+
+      state.value = mount.apply(this, arguments)
```

Comment:
Why not simply `mount()`?

Reply:
You're right, `mount` is an arrow function and, by definition, already inherits `this` and `arguments` from the decorated function.

```suggestion
      state.value = mount()
```